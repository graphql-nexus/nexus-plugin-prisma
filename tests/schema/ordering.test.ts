import { objectType } from 'nexus'
import { generateSchemaAndTypes } from '../__utils'

it('orderby fields can be allow-listed', async () => {
  const datamodel = `
    model Foo {
      id  Int @id @default(autoincrement())
      a   String
      b   String
      c   String
    }
  `

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [
    objectType({
      name: 'Foo',
      definition(t: any) {
        t.model.id()
      },
    }),
    objectType({
      name: 'Query',
      definition(t: any) {
        t.crud.foos({ ordering: { b: true } })
      },
    }),
  ])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})

it('finds correct ordering types if fullTextSearch preview feature is enabled', async () => {
  const datamodel = `
    generator prisma_client {
      provider = "prisma-client-js"
      output   = "../../node_modules/.prisma/client"
      previewFeatures = ["fullTextSearch"]
    }

    datasource db {
      provider = "postgres"
      url      = env("DB_URL")
    }

    model Foo {
      id    Int @id @default(autoincrement())
      a     String
      bars  Bar[]
    }

    model Bar {
      id    Int @id @default(autoincrement())
      fooId Int?
      foo   Foo? @relation(fields: [fooId], references: [id])
    }
  `

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [
    objectType({
      name: 'Foo',
      definition(t: any) {
        t.model.id()
        t.model.a()
        t.model.bars({ ordering: true })
      },
    }),
    objectType({
      name: 'Bar',
      definition(t: any) {
        t.model.id()
        t.model.foo()
      },
    }),
    objectType({
      name: 'Query',
      definition(t: any) {
        t.crud.foos({
          ordering: true,
        })
      },
    }),
  ])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})
