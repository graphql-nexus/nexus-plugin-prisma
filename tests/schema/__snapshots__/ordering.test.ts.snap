// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`finds correct ordering types if fullTextSearch preview feature is enabled: schema 1`] = `
"type Foo {
  id: Int!
  a: String!
  bars(orderBy: [BarOrderByWithRelationAndSearchRelevanceInput!], first: Int, last: Int, before: BarWhereUniqueInput, after: BarWhereUniqueInput): [Bar!]!
}

type Bar {
  id: Int!
  foo: Foo
}

type Query {
  foos(orderBy: [FooOrderByWithRelationAndSearchRelevanceInput!], first: Int, last: Int, before: FooWhereUniqueInput, after: FooWhereUniqueInput): [Foo!]!
}

input BarOrderByWithRelationAndSearchRelevanceInput {
  id: SortOrder
  fooId: SortOrderInput
  foo: FooOrderByWithRelationAndSearchRelevanceInput
}

input BarWhereUniqueInput {
  id: Int
  AND: [BarWhereInput!]
  OR: [BarWhereInput!]
  NOT: [BarWhereInput!]
  fooId: IntNullableFilter
  foo: FooWhereInput
}

input FooOrderByWithRelationAndSearchRelevanceInput {
  id: SortOrder
  a: SortOrder
  bars: BarOrderByRelationAggregateInput
  _relevance: FooOrderByRelevanceInput
}

input FooWhereUniqueInput {
  id: Int
  AND: [FooWhereInput!]
  OR: [FooWhereInput!]
  NOT: [FooWhereInput!]
  a: StringFilter
  bars: BarListRelationFilter
}

enum SortOrder {
  asc
  desc
}

input SortOrderInput {
  sort: SortOrder!
  nulls: NullsOrder
}

input BarWhereInput {
  AND: [BarWhereInput!]
  OR: [BarWhereInput!]
  NOT: [BarWhereInput!]
  id: IntFilter
  fooId: IntNullableFilter
  foo: FooWhereInput
}

input IntNullableFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntNullableFilter
}

input FooWhereInput {
  AND: [FooWhereInput!]
  OR: [FooWhereInput!]
  NOT: [FooWhereInput!]
  id: IntFilter
  a: StringFilter
  bars: BarListRelationFilter
}

input BarOrderByRelationAggregateInput {
  _count: SortOrder
}

input FooOrderByRelevanceInput {
  fields: FooOrderByRelevanceFieldEnum!
  sort: SortOrder!
  search: String!
}

input StringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  search: String
  mode: QueryMode
  not: NestedStringFilter
}

input BarListRelationFilter {
  every: BarWhereInput
  some: BarWhereInput
  none: BarWhereInput
}

enum NullsOrder {
  first
  last
}

input IntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input NestedIntNullableFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntNullableFilter
}

enum FooOrderByRelevanceFieldEnum {
  a
}

enum QueryMode {
  default
  insensitive
}

input NestedStringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  search: String
  not: NestedStringFilter
}

input NestedIntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}"
`;

exports[`finds correct ordering types if fullTextSearch preview feature is enabled: typegen 1`] = `
"import * as Typegen from '@morgothulhu/nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
    first?: boolean
    last?: boolean
    before?: boolean
    after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  Foo: Prisma.Foo
  Bar: Prisma.Bar
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    foos: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'a' | 'bars'
      ordering: 'id' | 'a' | 'bars' | '_relevance'
    }
    groupByFoo: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'a' | 'bars'
      ordering: 'id' | 'a' | '_count' | '_avg' | '_max' | '_min' | '_sum'
    }
    bars: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'fooId' | 'foo'
      ordering: 'id' | 'fooId' | 'foo'
    }
    groupByBar: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'fooId' | 'foo'
      ordering: 'id' | 'fooId' | '_count' | '_avg' | '_max' | '_min' | '_sum'
    }
  },
  Foo: {
    bars: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'fooId' | 'foo'
      ordering: 'id' | 'fooId' | 'foo'
    }
  }
  Bar: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    aggregateFoo: 'AggregateFoo'
    foos: 'Foo'
    foo: 'Foo'
    groupByFoo: 'FooGroupByOutputType'
    aggregateBar: 'AggregateBar'
    bars: 'Bar'
    bar: 'Bar'
    groupByBar: 'BarGroupByOutputType'
  },
  Mutation: {
    createOneFoo: 'Foo'
    createManyFoo: 'AffectedRowsOutput'
    deleteOneFoo: 'Foo'
    deleteManyFoo: 'AffectedRowsOutput'
    updateOneFoo: 'Foo'
    updateManyFoo: 'AffectedRowsOutput'
    upsertOneFoo: 'Foo'
    createOneBar: 'Bar'
    createManyBar: 'AffectedRowsOutput'
    deleteOneBar: 'Bar'
    deleteManyBar: 'AffectedRowsOutput'
    updateOneBar: 'Bar'
    updateManyBar: 'AffectedRowsOutput'
    upsertOneBar: 'Bar'
  },
  Foo: {
    id: 'Int'
    a: 'String'
    bars: 'Bar'
  }
  Bar: {
    id: 'Int'
    fooId: 'Int'
    foo: 'Foo'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  Foo: Typegen.NexusPrismaFields<'Foo'>
  Bar: Typegen.NexusPrismaFields<'Bar'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`orderby fields can be allow-listed: schema 1`] = `
"type Foo {
  id: Int!
}

type Query {
  foos(orderBy: [QueryFoosOrderByInput!], first: Int, last: Int, before: FooWhereUniqueInput, after: FooWhereUniqueInput): [Foo!]!
}

input QueryFoosOrderByInput {
  b: SortOrder
}

input FooWhereUniqueInput {
  id: Int
  AND: [FooWhereInput!]
  OR: [FooWhereInput!]
  NOT: [FooWhereInput!]
  a: StringFilter
  b: StringFilter
  c: StringFilter
}

enum SortOrder {
  asc
  desc
}

input FooWhereInput {
  AND: [FooWhereInput!]
  OR: [FooWhereInput!]
  NOT: [FooWhereInput!]
  id: IntFilter
  a: StringFilter
  b: StringFilter
  c: StringFilter
}

input StringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input IntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input NestedStringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input NestedIntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}"
`;

exports[`orderby fields can be allow-listed: typegen 1`] = `
"import * as Typegen from '@morgothulhu/nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
    first?: boolean
    last?: boolean
    before?: boolean
    after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  Foo: Prisma.Foo
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    foos: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'a' | 'b' | 'c'
      ordering: 'id' | 'a' | 'b' | 'c'
    }
    groupByFoo: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'a' | 'b' | 'c'
      ordering: 'id' | 'a' | 'b' | 'c' | '_count' | '_avg' | '_max' | '_min' | '_sum'
    }
  },
  Foo: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    aggregateFoo: 'AggregateFoo'
    foos: 'Foo'
    foo: 'Foo'
    groupByFoo: 'FooGroupByOutputType'
  },
  Mutation: {
    createOneFoo: 'Foo'
    deleteOneFoo: 'Foo'
    deleteManyFoo: 'AffectedRowsOutput'
    updateOneFoo: 'Foo'
    updateManyFoo: 'AffectedRowsOutput'
    upsertOneFoo: 'Foo'
  },
  Foo: {
    id: 'Int'
    a: 'String'
    b: 'String'
    c: 'String'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  Foo: Typegen.NexusPrismaFields<'Foo'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;
